plugin/evm/header/base_fee_test.go

```diff
@@ -11,328 +11,344 @@ import (
+					smoothingFactor            = ap3.BaseFeeChangeDenominator
+					baseFeeFractionUnderTarget = amountUnderTarget * parentBaseFee / gasTarget
+					delta                      = baseFeeFractionUnderTarget / smoothingFactor
+					baseFee                    = parentBaseFee - delta
+				)
+				return big.NewInt(baseFee)
 			}(),
 		},
-		// Test overflow handling
 		{
-			minFee: big.NewInt(ap3.MinBaseFee),
-			maxFee: big.NewInt(ap3.MaxBaseFee),
-			blocks: func() []blockDefinition {
-				blocks := make([]blockDefinition, 0, len(spacedTimestamps))
-				for _, timestamp := range spacedTimestamps {
-					blocks = append(blocks, blockDefinition{
-						timestamp: timestamp,
-						gasUsed:   math.MaxUint64,
-					})
-				}
-				return blocks
+			name:     "ap3_large_decrease",
+			upgrades: params.TestApricotPhase3Config.NetworkUpgrades,
+			parent: &types.Header{
+				Number:  big.NewInt(1),
+				Extra:   feeWindowBytes(ap3.Window{}),
+				BaseFee: big.NewInt(ap3.MaxBaseFee),
+			},
+			timestamp: 2 * ap3.WindowLen,
+			want: func() *big.Int {
+				const (
+					gasTarget                  = ap3.TargetGas
+					gasUsed                    = 0
+					amountUnderTarget          = gasTarget - gasUsed
+					parentBaseFee              = ap3.MaxBaseFee
+					smoothingFactor            = ap3.BaseFeeChangeDenominator
+					baseFeeFractionUnderTarget = amountUnderTarget * parentBaseFee / gasTarget
+					windowsElapsed             = 2
+					delta                      = windowsElapsed * baseFeeFractionUnderTarget / smoothingFactor
+					baseFee                    = parentBaseFee - delta
+				)
+				return big.NewInt(baseFee)
 			}(),
 		},
 		{
-			minFee: big.NewInt(ap3.MinBaseFee),
-			maxFee: big.NewInt(ap3.MaxBaseFee),
-			blocks: []blockDefinition{
-				{
-					timestamp: 1,
-					gasUsed:   1_000_000,
-				},
-				{
-					timestamp: 3,
-					gasUsed:   1_000_000,
-				},
-				{
-					timestamp: 5,
-					gasUsed:   2_000_000,
-				},
-				{
-					timestamp: 5,
-					gasUsed:   6_000_000,
-				},
-				{
-					timestamp: 7,
-					gasUsed:   6_000_000,
-				},
-				{
-					timestamp: 1000,
-					gasUsed:   6_000_000,
-				},
-				{
-					timestamp: 1001,
-					gasUsed:   6_000_000,
-				},
-				{
-					timestamp: 1002,
-					gasUsed:   6_000_000,
-				},
+			name:     "ap3_increase",
+			upgrades: params.TestApricotPhase3Config.NetworkUpgrades,
+			parent: &types.Header{
+				Number:  big.NewInt(1),
+				GasUsed: 2 * ap3.TargetGas,
+				Extra:   feeWindowBytes(ap3.Window{}),
+				BaseFee: big.NewInt(ap3.MinBaseFee),
 			},
+			timestamp: 1,
+			want: func() *big.Int {
+				const (
+					gasTarget                  = ap3.TargetGas
+					gasUsed                    = 2*ap3.TargetGas + ap3.IntrinsicBlockGas
+					amountOverTarget           = gasUsed - gasTarget
+					parentBaseFee              = ap3.MinBaseFee
+					smoothingFactor            = ap3.BaseFeeChangeDenominator
+					baseFeeFractionUnderTarget = amountOverTarget * parentBaseFee / gasTarget
```

Comment:
should this be Over?
I understand it's for clarity but I think the calculation from this point can be done in a (local) function like baseFee(amount, target, parentBaseFee, smoothingFactor)
on the other hand, it's nice that one can mouseover and see the value.. not sure.

Reply:
🦅 👁️

Reply:
I did it like this so that we could see the value of each constant in the IDE... but 🤷